### PR TITLE
[DRAFT] RFC: Associated type defaults and default groups

### PR DESCRIPTION
**Pull request by Centril**
This is a **draft** version of an RFC for you to review, before a formal proposal is made for consideration.

# Summary

[RFC 192]: https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md#defaults

1. Resolve the design of associated type defaults, first introduced in [RFC 192], such that provided methods and other items may not assume type defaults. This applies equally to `default` with respect to specialization.

2. Introduce the concept of `default { .. }` groups in traits and their implementations which may be used to introduce atomic units of specialization (if anything in the group is specialized, everything must be).

### CODE REVIEW

File: text/0000-assoc-default-groups.md [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+[the section where it is discussed][default_groups].
+
+# Prior art
+[prior-art]: #prior-art
+
+## Haskell
+
+As Rust traits are a form of type classes,
+we naturally look for prior art from were they first were introduced.
+That language, being Haskell, permits a user to specify associated type defaults.
```
Comment (kennytm) [inline text/0000-assoc-default-groups.md]: While traits are type classes, it is also comparable to OOP interfaces; [specialization](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md) in particular is entirely inspired from C++.

So I'd suggest adding the following comparison to prior arts as well:

* C++ (all diagnostic is done post-monomorphization so probably not an interesting example, but still...)
* Swift (Rust trait ≈ Swift protocols. Protocols can be inherited and can have associated types. But protocol constants and methods cannot be defaulted.)
* Perhaps also check what Scala is doing 🙃

File: text/0000-assoc-default-groups.md [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+
+   ```rust
+   trait Arbitrary: Sized + fmt::Debug {
+       type Parameters: Default = ();
+   
+       fn arbitrary_with(args: Self::Parameters) -> Self::Strategy;
+
+       fn arbitrary() -> Self::Strategy {
+           Self::arbitrary_with(Default::default())
+       }
```
Comment (scottmcm) [inline text/0000-assoc-default-groups.md:55]: This might be a side point, but when I look at this I think `arbitrary()` should have the `where Self::Parameters: Default` bound, and it not be a bound on the associated type.

Comment (Centril) [reply text/0000-assoc-default-groups.md:55]: Modulo implied bounds, this is equivalent semantically. However, since implied bounds are yet to be stabilized, you can't use `where Self::Parameters: Default` without giving up on ergonomics; afaik.
See https://github.com/rust-lang/rfcs/blob/master/text/2289-associated-type-bounds.md#an-alternative-desugaring-of-bounds-on-associated-types for a discussion.

File: text/0000-assoc-default-groups.md [hunk 3]
```diff
@@ -0,0 +1,957 @@
...
+   the grammar of `impl`s. In addition, this brings `trait`s more in line
+   with `type` aliases.
+
+The following points were also noted in [RFC 192], but we expand upon them here:
+
+3. Most notably, type defaults allow you to provide more ergonomic APIs.
+
+   [proptest]: https://altsysrq.github.io/rustdoc/proptest/latest/proptest/arbitrary/trait.Arbitrary.html
+
+   For example, we may provide an API (due to [proptest]):
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: It might read better to write this line as:

`For example, we could change [proptest]'s API to be:`

File: text/0000-assoc-default-groups.md [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+
+    default fn quux(x: Self::Bar) -> u8 { x }
+}
+```
+
+The compiler will reject this because you are not allowed to assume,
+just like before, that `x: u8`. The reason why is much the same as
+we have previously discussed in the [background].
+
+Once place where this proposal does diverge from what is currently implemented
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: Once -> One; does diverge -> diverges

File: text/0000-assoc-default-groups.md [hunk 5]
```diff
@@ -0,0 +1,957 @@
...
+
+## `default` specialization groups
+[default_groups]: #default-specialization-groups
+
+Now, you might be thinking: - *"Well, what if I __do__ need to assume that
+my defaulted associated type is what I said in a provided method,
+what do I do then?"*. Don't worry; We've got you covered.
+
+To be able to assume that `Self::Bar` is truly `u8` in snippets (2) and (5),
+you may henceforth use `default { .. }` to group items into atomic units of
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: Might want to explicitly state `associated items` here.

File: text/0000-assoc-default-groups.md [hunk 6]
```diff
@@ -0,0 +1,957 @@
...
+```rust
+trait ComputerScientist {
+    default {
+        type Bar = u8;
+        const QUUX: Self::Bar = 42u8; // OK!
+        fn wibble(x: Self::Bar) -> u8 { x } // OK!
+    }
+}
+
+struct Alan;
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: At RustConf, misdreavious's slides included descriptions of what was being referred to. It was a great thing to include. I feel like we should do the same for RFCs, and link to these peoples' Wikipedia pages.

Comment (Havvy) [inline text/0000-assoc-default-groups.md]: Can we change two of these to be women who've contributed to computer science, to maintain a gender balance?

File: text/0000-assoc-default-groups.md [hunk 7]
```diff
@@ -0,0 +1,957 @@
...
+}
+
+impl<T> Foo for Vec<T> {
+    fn baz(x: Self::Bar) -> usize { x }
+}
+```
+
+In the current implementation, (6) is rejected because the compiler will not
+let you assume that `x` is of type `usize`. But in this proposal, you would be
+allowed to assume this. To permit this is not a problem because `Foo for ()`
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: `()` => `Vec<T>`

File: text/0000-assoc-default-groups.md [hunk 8]
```diff
@@ -0,0 +1,957 @@
...
+preserves parametricity and lacks specialization,
+wherefore `default { .. }` might not carry its weight.
+
+## Idris
+
+[idris_interface]: http://docs.idris-lang.org/en/latest/tutorial/interfaces.html
+
+Idris has a concept it calls [`interface`s][idris_interface].
+These resemble type classes in Haskell, and by extension traits in Rust.
+However, unlike Haskell and Rust, these `interface`s are incoherent and will
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: Incoherent is a strongly negative word. Would be good to rephrase this to be less negative?

File: text/0000-assoc-default-groups.md [hunk 9]
```diff
@@ -0,0 +1,957 @@
...
+
+To be able to assume that `Self::Bar` is truly `u8` in snippets (2) and (5),
+you may henceforth use `default { .. }` to group items into atomic units of
+specialization. This means that if one item in `default { .. }` is overridden
+in an implementation, then all all the items must be. An example (7):
+
+```rust
+trait ComputerScientist {
+    default {
+        type Bar = u8;
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: "Bar, QUUX, wibble" keep confusing me. Could these be changed to terms in the domain of being a Computer Scientist?

File: text/0000-assoc-default-groups.md [hunk 10]
```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
...
+}
+```
+
+However, we may not write:
+
+```swift
+protocol Foo {
+    associatedtype Bar = Int
+
+    func append() -> Bar { return 0 }
```
Comment (strega-nil) [inline text/0000-assoc-default-groups.md:1026]: The syntax highlighter seems to dislike this, maybe split it?

File: text/0000-assoc-default-groups.md [hunk 11]
```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
...
+
+```cc
+main.cpp: In function 'int main()':
+main.cpp:21:24: error: 'struct foo<wrap<void> >' has no member named 'make_a_bar'
+     std::cout << b_foo.make_a_bar() << std::endl;
+                        ^~~~~~~~~~
+```
+
+This shows that templates in C++ are fundamentally different from the type of
+parametric polymorphism (generics) that Rust employs.
```
Comment (strega-nil) [inline text/0000-assoc-default-groups.md]: I would cut out this part, from the "do note" on, and rewrite the above paragraph.

File: text/0000-assoc-default-groups.md [hunk 12]
```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
...
+
+    foo<wrap<void>> b_foo;
+    std::cout << b_foo.make_a_bar() << std::endl;
+}
+```
+
+One thing to note here is that C++ allows us to assume in both the unspecialized
+variant of `foo` as well as the specialized version that `bar` is the underlying
+type we said it was in the `typedef`. This is unlike the default in this RFC
+but the same as when we use `default { .. }`.
```
Comment (strega-nil) [inline text/0000-assoc-default-groups.md]: I'd rewrite this comment quite heavily, to something along the following lines:

> You will note that C++ allows us to assume in both the base template class, as well as the specialization, that `bar` is equal to the underlying type. This is because one cannot specialize any part of a class without specializing the whole of it. It's equivalent to one atomic `default { ... }` block.

File: text/0000-assoc-default-groups.md [hunk 13]
```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
...
+template<typename T> struct wrap {};
+
+template<typename T> struct foo { // Unspecialized.
+    typedef int bar;
+
+    bar make_a_bar() { return 0; };
+};
+
+template<typename T> struct foo<wrap<T>> { // Partial specialization.
+    typedef std::string bar;
```
Comment (strega-nil) [inline text/0000-assoc-default-groups.md]: You should use `using` here instead - `using bar = std::string;`. It's just more readable.

File: text/0000-assoc-default-groups.md [hunk 14]
```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
...
+This is shown in the following example:
+
+```cpp
+#include <iostream>
+#include <string>
+
+template<typename T> struct wrap {};
+
+template<typename T> struct foo { // Unspecialized.
+    typedef int bar;
```
Comment (strega-nil) [inline text/0000-assoc-default-groups.md]: See above comment on `using`.

**Review by strega-nil — COMMENTED** [review]
I like the base RFC, mostly editorial stuff, plus some corrections on the C++.

File: text/0000-assoc-default-groups.md [hunk 15]
```diff
@@ -0,0 +1,957 @@
...
+    }
+
+    fn foo() {
+        let _: Self::Details = 1.0f32;
+        // ERROR! Can't assume it is f32.
+    }
+}
+```
+
+However, please note that for use of `default { .. }` inside implementations,
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: This is worth having at the top of the section, not the bottom.

File: text/0000-assoc-default-groups.md [hunk 16]
```diff
@@ -0,0 +1,957 @@
...
+
+The main drawbacks of this proposal are that:
+
+1. `default { .. }` is introduced, adding to the complexity of the language.
+
+   However, it should be noted that token `default` is already accepted for
+   use by specialization and for `default impl`.
+   Therefore, the syntax is only partially new.
+
+2. secondarily, if you have implementations where it is commonly needed
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md]: "secondarily" is already said by the number "2".

File: text/0000-assoc-default-groups.md (hunk 10 above)
Comment (Havvy) [reply text/0000-assoc-default-groups.md:1026]: The syntax highlighter dislikes it because it's literally not allowed.

Comment (strega-nil) [reply text/0000-assoc-default-groups.md:1026]: Ohhh, okay

**Comment by Havvy** [comment]
I have an alternative in mind.

So, a caveat here. I have not thought about how this interacts with specialization. So everything here talks from the perspective of writing defaults in traits.

The short of it is that in this alternative, default blocks let you specialize default implementations when the implementing type's type aliases match the specified type aliases in the default block. And furthermore, the default block does not allow declaration of new associated items. It can only provide new defaults that can make use of the specified type.

For an example of this alternative, we'll use this trait and this impl:

```rust
trait Trait {
  type Type;
  const Const: Type;
}

impl Trait for UStruct {
  type Type = u8;
  const Const: Type = 0u8;
}
```

Without using a default block, we cannot give a default to `Const` for the same reasons the RFC already spells out. But we want to have it so that when `Type` is `u8`, then `Const` is `0u8`. So to do so, we add a default block to the trait.

```rust
trait Trait {
  type Type;
  const Const: Type;

  default {
    type Type = u8;
    const Const: u8 = 0u8;
  }
}

impl Trait for UStruct {
  type Type = u8;
  // No need to specify `Const` as it uses the default given.
}
```

But then, we decide that if the implementer chooses `i8` for their type, we should also have a different default constant for them. To fulfill this, we just add a new default block.

```rust
trait Trait {
  type Type;
  const Const: Type;

  default {
    type Type = u8;
    const Const: u8 = 0u8;
  }

  default {
    type Type = i8;
    const Const: i8 = 1i8;
  }
}

impl Trait for IStruct {
  type Type = i8;
  // `Const` defaults to 1i8
}
```

To pick out which default applies, we look at what the `type`s are in the default block compared to what they are in the implementation. And since this is about allowing default on type aliases outside of the default blocks, if it's there, and the implementation doesn't override, then that's the type. And if there's a default block for that, it also uses those as defaults. So, let's give `Trait` a default `Type` of `u8`.

```rust
trait Trait {
  type Type = u8;
  const Const: Type;

  default {
    type Type = u8;
    const Const: u8 = 0u8;
  }

  default {
    type Type = i8;
    const Const: i8 = 1i8;
  }
}

impl Trait for UStruct {
  // type defaults to `u8`.
  // `Const` defaults to 0u8 because `Type` is `u8`.
}
```

And this all works great in simple cases. It gets complex though, if there's more than one type. You could have...

```rust
trait Trait {
  type TypeA;
  type TypeB;
  const Const: u8;

  default {
    type TypeA = ();
    const Const: u8= 0u8;
  }

  default {
    type TypeB = ();
    const Const: u8= 1u8;
  }
}
```

...and then you're not sure which default applies. That can be fixed by requiring defaulted items use all type aliases in the default block. Similarly, you can have specialization issues if you allow defaults where the types are generic.

**Comment by kennytm** [comment]
In RFC #2500 I've got this trait:

```rust
trait Needle<H: Haystack>: Sized {
    type Searcher: Searcher<H::Target>;
    type Consumer: Consumer<H::Target>;
    fn into_searcher(self) -> Self::Searcher;
    fn into_consumer(self) -> Self::Consumer;
}
```

usually the associated types Consumer and Searcher are equivalent, so I'd like to default it:

```rust
trait Needle<H: Haystack>: Sized {
    type Searcher: Searcher<H::Target>;
    fn into_searcher(self) -> Self::Searcher;

    default {
        type Consumer: Consumer<H::Target> = Self::Searcher;
        fn into_consumer(self) -> Self::Consumer { self.into_searcher() }
    }
}
```

However, a Searcher may not implement `Consume<H::Target>`, and the default should not be applied in this case (should cause error). Could this RFC handle this?

**Comment by varkor** [comment]
I haven't looked at others' comments, but this seems like a nice, straightforward approach to defaults to me.
I would like to see `impl Trait` mentioned here, but it's not passed an RFC yet so it's not technically relevant now. But specifically my question is that if an associated type default is provided:
```rust
type Foo = impl Bar;
```
is the opaque type that implements `Bar` here shared between all implementors of the trait, or re-inferred for all implementers of the trait? Or is this construction entirely banned for now?

This could be left unaddressed unless the trait alias `impl Trait` is accepted, but I think it's worth bearing in mind.

File: text/0000-assoc-default-groups.md [hunk 17]
```diff
@@ -0,0 +1,1078 @@
...
+   ```rust
+   trait Arbitrary: Sized + fmt::Debug {
+       fn arbitrary() -> Self::Strategy;
+   
+       type Strategy: Strategy<Value = Self>;
+   }
+   ```
+
+   By allowing defaults, we can transition to this more flexible API without
+   breaking any consumers by simply saying:
```
Comment (alexreg) [inline text/0000-assoc-default-groups.md:93]: Maybe just clarify by indicating consumers == `impl Arbitrary` for some type?

File: text/0000-assoc-default-groups.md [hunk 18]
```diff
@@ -0,0 +1,1078 @@
...
+### Changes in this RFC
+[changes]: #changes-in-this-rfc
+
+In this RFC, we change the approach in [RFC 192] to the currently implemented
+approach. Thus, you will continue to receive the error message above
+and you will be able to provide associated type defaults.
+
+[specialization]: https://github.com/rust-lang/rfcs/pull/1210
+
+With respect to [specialization], the behaviour is the same.
```
Comment (alexreg) [inline text/0000-assoc-default-groups.md:277]: What what, mixed US/British English?! 😯

File: text/0000-assoc-default-groups.md [hunk 19]
```diff
@@ -0,0 +1,1078 @@
...
+
+The compiler will reject this because you are not allowed to assume,
+just like before, that `x: u8`. The reason why is much the same as
+we have previously discussed in the [background].
+
+One place where this proposal diverges from what is currently implemented
+is with respect to the following example (6):
+
+```rust
+#![feature(associated_type_defaults)]
```
Comment (alexreg) [inline text/0000-assoc-default-groups.md]: Slightly confusing feature name, given we already have "associated type defaults"? Maybe "associated_type_defaults_v2" or such?

File: text/0000-assoc-default-groups.md [hunk 20]
```diff
@@ -0,0 +1,1078 @@
...
+trait Foo {
+    type Assoc = impl Bar;
+
+    ...
+}
+```
+
+the underlying type of `Assoc` stays the same for all implementations which
+do not change the default of `Assoc`. The same applies to specializations.
+With respect to type visibility, it is the same as that of `existential type`.
```
Comment (alexreg) [inline text/0000-assoc-default-groups.md]: "Type opacity" (rather than "visibility"), you mean? I think that's clearer.

File: text/0000-assoc-default-groups.md [hunk 21]
```diff
@@ -0,0 +1,1078 @@
...
+*all* items must be overridden in that group.
+
+Items inside a group may assume the definitions inside the group.
+Items outside of that group may not assume the definitions inside of it.
+
+#### Nesting
+
+There applies no restriction on the nesting of groups.
+This means that you may nest them arbitrarily.
+When nesting does occur, the atomicity applies as if the nesting was flattened.
```
Comment (alexreg) [inline text/0000-assoc-default-groups.md]: s/was flattened/were flattened/.

Wouldn't it be better just to disallow nesting? I thought of this before I read this section, and came to the conclusion that such variations just add needless complexity and potentially confusion.

**Review by alexreg — COMMENTED** [review]
Nice work. Overall, definitely in favour of this RFC. I must say, I didn't fully appreciate the motivation until I read the Guide section (not so much the Motivation section, hah). I was weighing up the pros and cons of the behaviour of associated type defaults in provided methods in this RFC vs. existing behaviour, and came to essentially the same conclusions as you. The granularity of this approach offers a big advantage, and the way methods interact with default associated types just seems more intuitive this way.

File: text/0000-assoc-default-groups.md [hunk 22]
```diff
@@ -0,0 +1,1104 @@
...
+
+With these changes,
+we consider the design of associated type defaults to be *finalized*.
+
+## `default` specialization groups
+[default_groups]: #default-specialization-groups
+
+Note: Everything in this section assumes actual support for [specialization].
+
+Now, you might be thinking: - *"Well, what if I __do__ need to assume that
```
Comment (Havvy) [inline text/0000-assoc-default-groups.md:335]: This question made me think that `default` specialization groups is only useful if said group contains an associated type. This is then never challenged by having all examples have associated types in them.

**Comment by Centril** [comment]
Thanks all for chipping in!
Time to publish :)
